### PR TITLE
Lashemilt/issue157

### PR DESCRIPTION
**Pull request by LAShemilt**
This seems to be easier than first thought. Amended the decorator to add the metadata from files to a directory. Wrote a new operation call Dependent Dummy Operation to test this.

### CODE REVIEW

File: rfi_file_monitor/file.py [hunk 1]
```diff
@@ -230,6 +230,7 @@ def __init__(self,
         self._filelist : List[Tuple[str, int]] = []
         self._filelist_timestamp : int = 0
         self._total_size : int = 0
+        self._total_size : int = 0
```
Comment (tschoonj) [inline rfi_file_monitor/file.py:233]: Remove this line

File: rfi_file_monitor/operations/dummy_operation.py [hunk 2]
```diff
@@ -129,9 +135,13 @@ def preflight_check(self):
...
             return add_directory_support(type(self)._run)(self, file)
-        return self._run(file)
+        return self._run(file, self.index)
+
+    @classmethod
+    def _attach_metadata(cls, file: File, operation_index:int):
+        file.operation_metadata[operation_index] = {'number':random.random()}
 
-    def _run(self, file: File):
+    def _run(self, file: File, operation_index):
```
Comment (tschoonj) [inline rfi_file_monitor/operations/dummy_operation.py:144]: `_run` is an object method, so you can call `self.index` from within it, no need to pass it as an argument.

File: rfi_file_monitor/utils/decorators.py [hunk 3]
```diff
@@ -133,14 +133,16 @@ def wrapper(self: Operation, file: File):
 
                 # run the wrapped method, and do the usual exception and return value handling
                 try:
-                    rv = run(self, _file)
+                    metadata, rv = run(self, _file)
```
Comment (tschoonj) [inline rfi_file_monitor/utils/decorators.py:136]: This is wrong, operation `run` methods return a single value.

File: rfi_file_monitor/utils/decorators.py [hunk 4]
```diff
@@ -133,14 +133,16 @@ def wrapper(self: Operation, file: File):
...
                 except SkippedOperation:
                     continue
                 # other exceptions should propagate
 
+                if _file.operation_metadata:
+                    file.operation_metadata[_file.filename] = _file.operation_metadata.values()
```
Comment (tschoonj) [inline rfi_file_monitor/utils/decorators.py:142]: `operation_metadata` is a dict with keys the numbers of the operation indices that attached data.

So you will need to do something like:

```python
file.operation_metadata[self.index][_file.filename] = _file.operation_metadata[self.index]
```

Ensure that before the loop over the files starts, that a dict is assigned to `file.operation_metadata[self.index]`.

**Review by tschoonj — COMMENTED** [review]
Hey Laura,

My main comment would be not to create a new dummy operation for this, and instead adding an option to the current DummyOperation to allow attaching operation metadata with random values. If you also change its run method to print the `operation_metadata` from previous steps, you will accomplish the same thing, if you simply put two dummy operations in a pipeline.

**Comment by LAShemilt** [comment]
> 
> 
> Hey Laura,
> 
> My main comment would be not to create a new dummy operation for this, and instead adding an option to the current DummyOperation to allow attaching operation metadata with random values. If you also change its run method to print the `operation_metadata` from previous steps, you will accomplish the same thing, if you simply put two dummy operations in a pipeline.

Hi Tom,
My main thought was to emulate the behaviour of Scicat by having one operation outside the add_directory_support decorator, to see if I could access metadata outside the dummy operation. If it is too cumbersome to have two methods, I can just do as you suggest.

**Pull request closed without merging**